### PR TITLE
refactor(frontend): modify Pin Snapshot in hummock_snapshot_manager.rs

### PR DESCRIPTION
**Pull request by ZENOTME**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

This PR modifies the Pin Snapshot method in hummock_snapshot_manager.rs.

In the old version, we check if the epoch is outdated to determine whether call pin_snapshot when we want to get an epoch.
For now, we create a background thread to pin_snapshot automatically.  When we want an epoch, we can directly read the last_epoch and needn't do other extra work. 

This method is refer to [local_version_manager.rs/{pin_version_with_retry, start_pin_worker}](https://github.com/singularity-data/risingwave/blob/main/src/storage/src/hummock/local_version_manager.rs#:~:text=%7D-,///%20Pin%20a%20version%20with%20retry.,%7D,-async%20fn%20start_unpin_worker)

## Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)
#2364

### CODE REVIEW

**Comment by ZENOTME** [comment]
A weird bug comes out. I execute the following statement in e2e_test and it failed.
```
statement ok
create table t1 (v1 int not null, v2 int not null) with ('appendonly' = true);

statement ok
insert into t1 values (1,2), (2,3);

query I
select * from t1;
----
1 2 
2 3
```

the error message is 
image

then I try this statement in psql, but the result is true.
dev => select * from t1;
v1 | v2
---+---
 1  |  2
 2  |  3

can you guess what error possible cause this bug? @BowenXiao1999

**Comment by BowenXiao1999** [comment]
> A weird bug comes out. I execute the following statement in e2e_test and it failed.
> 
> ```
> statement ok
> create table t1 (v1 int not null, v2 int not null) with ('appendonly' = true);
> 
> statement ok
> insert into t1 values (1,2), (2,3);
> 
> query I
> select * from t1;
> ----
> 1 2 
> 2 3
> ```
> 
> the error message is image
> 
> then I try this statement in psql, but the result is true. dev => select * from t1; v1 | v2 ---+--- 1 | 2 2 | 3
> 
> can you guess what error possible cause this bug? @BowenXiao1999

Pull main and try again. This error only happen in CI?

**Comment by ZENOTME** [comment]
> Pull main and try again. This error only happen in CI?

It doesn't work. I test locally. It looks like caused by this PR.

**Comment by BowenXiao1999** [comment]
OK, I'll pull this branch and try it locally.

File: src/frontend/src/observer/observer_manager.rs [hunk 1]
```diff
@@ -204,9 +204,10 @@ impl ObserverManager {
                     resp
                 )
             }
-            Info::HummockSnapshot(hummock_snapshot) => {
-                self.hummock_snapshot_manager
-                    .update_snapshot_status(hummock_snapshot.epoch)
+            Info::HummockSnapshot(_hummock_snapshot) => {
+                // NOTE:DO NOTHING,FOR NOW WE DONT USE HUMMOCK SNAPSHOT
+                self._hummock_snapshot_manager
+                    .update_snapshot_status(_hummock_snapshot.epoch)
```
Comment (BowenXiao1999) [inline src/frontend/src/observer/observer_manager.rs:210]: revert the change?

File: src/frontend/src/scheduler/error.rs [hunk 2]
```diff
@@ -16,12 +16,10 @@ use risingwave_common::error::{ErrorCode, RwError, TrackingIssue};
...
 use thiserror::Error;
 
-use crate::scheduler::plan_fragmenter::QueryId;
-
 #[derive(Error, Debug)]
 pub enum SchedulerError {
-    #[error("Pin snapshot error: {0} fails to get epoch {1}")]
-    PinSnapshot(QueryId, u64),
+    #[error("Pin snapshot error: fails to get epoch {0}, max retry reahced {1}")]
+    PinSnapshot(u64, u64),
```
Comment (BowenXiao1999) [inline src/frontend/src/scheduler/error.rs:22]: Why change this?

File: src/frontend/src/scheduler/hummock_snapshot_manager.rs [hunk 3]
```diff
@@ -31,25 +34,32 @@ pub struct HummockSnapshotManager {
...
+            .expect("should be able to pinned the first epoch");
+
+        // Create the snapshot manager.
+        let local_snapshot_manager = Arc::new(Self {
+            core: Mutex::new(HummockSnapshotManagerCore::new(epoch)),
             meta_client,
-        }
+        });
+
+        // Pin and get the latest version.
```
Comment (BowenXiao1999) [inline src/frontend/src/scheduler/hummock_snapshot_manager.rs:53]:
```suggestion
        // Pin the latest version periodically.
```

File: src/frontend/src/scheduler/hummock_snapshot_manager.rs [hunk 4]
```diff
@@ -67,16 +77,27 @@ impl HummockSnapshotManager {
...
-                    core_guard.is_outdated = true;
+                // Updated the outdated epoch.
+                if query_ids.is_empty() && core_guard.last_pinned == epoch {
+                    let epoch = Self::pin_epoch_with_retry(
+                        self.meta_client.clone(),
+                        Default::default(),
+                        usize::MAX,
+                        || false,
+                    )
+                    .await
```
Comment (BowenXiao1999) [inline src/frontend/src/scheduler/hummock_snapshot_manager.rs:88]: The question is: This PR still not differ too much from previous: pin_snapshot still in the critical path of query.

File: src/frontend/src/scheduler/hummock_snapshot_manager.rs [hunk 5]
```diff
@@ -108,30 +129,131 @@ impl HummockSnapshotManager {
...
+            };
+        }
+    }
+
     /// Used in `ObserverManager`.
     pub async fn update_snapshot_status(&self, epoch: u64) {
         let mut core_guard = self.core.lock().await;
         if core_guard.last_pinned < epoch {
-            core_guard.is_outdated = true;
+            let epoch = Self::pin_epoch_with_retry(
```
Comment (BowenXiao1999) [inline src/frontend/src/scheduler/hummock_snapshot_manager.rs:229]: This part of code seems can be reused.

File: src/frontend/src/scheduler/hummock_snapshot_manager.rs [hunk 6]
```diff
@@ -108,30 +129,131 @@ impl HummockSnapshotManager {
...
 
 impl HummockSnapshotManagerCore {
-    fn new() -> Self {
+    fn new(last_pinned: u64) -> Self {
         Self {
-            // Initialize by setting `is_outdated` to `true`.
-            is_outdated: true,
-            ..Default::default()
+            last_pinned,
+            epoch_to_query_ids: BTreeMap::from([(last_pinned, HashSet::new())]),
```
Comment (BowenXiao1999) [inline src/frontend/src/scheduler/hummock_snapshot_manager.rs:256]: Actually the last_pinned can be removed, there is retryable problems in quip: https://singularity-data.quip.com/Qbv8AypJHU68/Retryable-pin-RPC-in-hummock after we introduce unpin_snapshot_before

**Comment by BowenXiao1999** [comment]
The biggest problem is still we can not remove the RPC from batch query execution path.

File: src/frontend/src/scheduler/error.rs (hunk 2 above)
Comment (ZENOTME) [reply src/frontend/src/scheduler/error.rs:22]: The original is PinSnapshot(QueryID,epoch)
We don't need it now because when we Pin snapshot we don't know the QueryID. 
I change it and reuse it in pin_epoch_with_retry()

File: src/frontend/src/scheduler/hummock_snapshot_manager.rs (hunk 4 above)
Comment (ZENOTME) [reply src/frontend/src/scheduler/hummock_snapshot_manager.rs:88]: For now, pin_snapshot is not in the get_epoch, but it is in unpin_snapshot.
I prepare to a new PR to wrap the epoch in an object to implement RAII and then change the unpin_snapshot method to avoid pin_snapshot in it.

Comment (BowenXiao1999) [reply src/frontend/src/scheduler/hummock_snapshot_manager.rs:88]: How to avoid pin_snapshot RFC and how this new design preserve the correctness? 

I am little confused about the flow now. Can you describe it more clearly or just use a diagram? https://sequencediagram.org/

Comment (ZENOTME) [reply src/frontend/src/scheduler/hummock_snapshot_manager.rs:88]: image

Oh..It looks like we don't need a background thread

Comment (BowenXiao1999) [reply src/frontend/src/scheduler/hummock_snapshot_manager.rs:88]: Elaborated more on the first event drive: in what situation why we need this update (when query ids of last_pinned is empty)?

Comment (ZENOTME) [reply src/frontend/src/scheduler/hummock_snapshot_manager.rs:88]:
```
    pub async fn unpin_snapshot(&self, epoch: u64, query_id: &QueryId) -> SchedulerResult<()> {
        let min_epoch = async {
            // Decrease the ref count of corresponding epoch. If all last pinned epoch is dropped,
            // mark as outdated.
            let mut core_guard = self.core.lock().await;
            let query_ids = core_guard.epoch_to_query_ids.get_mut(&epoch);
            if let Some(query_ids) = query_ids {
                query_ids.remove(query_id);
                // Updated the outdated epoch.
                if query_ids.is_empty() && core_guard.last_pinned == epoch {                                <==============Here
                    Self::try_update_last_pinned(&mut core_guard, self.meta_client.clone()).await;    <==============Here
                }
            }
```
In this case, we need an update.
I try to comment the update and it cause the previous bug. It works when I add update here.
So I guess:In this case, we unpin the last query in last_pin, and this last_pin is outdated and we need to pin_snapshot to update it.

**Comment by ZENOTME** [comment]
consistency problem solved by #3557. Push-based need new design.
